Content addressability

tonistiigi opened the pull request:
This PR changes the Docker engine to use content addressable storage for images and layers. This means that image IDs are no longer arbitrarily assigned - now they are computed based on the filesystem contents and the image configuration. Images can securely share their underlying layers without duplicating data on disk.

This work involves a big refactoring of image-related code in the engine. The `graph` package is entirely removed, and is replaced by a few new packages:
- `layer`: Layer store interface and implementation. This manages filesystem layers without being aware of image configurations. The layer store maintains reference counts for layers and removes unreferenced layers.
- `image`: Image store interface and implementation. This manages image configurations. Image configurations include runtime configuration and reference the underlying filesystem data.
- `tag`: The tag functionality from the `graph` package has been separated out and moved here. Tags now use types from the `distribution/reference` package instead of plain strings.
- `distribution`: Push and pull code that originally was part of the `graph` package has been separated out into its own package.

The first time a version of the engine with these commits is started, it will migrate old graph metadata to the new format. This involves calculating content hashes for the existing data, but it does not move underlying graphdriver filesystem data. It doesn’t remove old graph metadata, so the migration process is not destructive.

The new data model does not have a one-to-one relationship between images and layers. A single image can have many layers. Existing versions of Docker create an image for each layer, and use the parent chain to link them together. That means that pulling a specific image requires pulling all the artifacts from the original build process of that image. With this PR, when an image is pulled from a registry, only a single image is created, to match the new data model. The history is preserved through a list of commands, dates, etc.
#### Summary of UI Changes:
- Pull/push do not transfer the entire parent chain.
- Full-length image IDs have a `sha256:` prefix. This prefix is hidden for truncated image IDs for convenience.
- Images don't have the concept of VirtualSize (because there’s no special meaning for a top layer anymore).
#### Future work:
- Add support for new manifest format for push/pull: https://github.com/docker/distribution/pull/1068
- Move graphdriver initialization/shutdown out of daemon.
- Move some layer, image, and distribution code out of docker/docker (remove dependencies on Engine code as necessary).
- Additional changes needed for managing the Windows base layer.
#### Unit test coverage for the new packages:

```
ok      github.com/docker/docker/image  0.051s  coverage: 85.8% of statements
ok      github.com/docker/docker/layer  0.223s  coverage: 76.0% of statements
ok      github.com/docker/docker/migrate/v1 0.033s  coverage: 71.1% of statements
ok      github.com/docker/docker/tag    0.017s  coverage: 86.1% of statements
```

For the full design document, see https://gist.github.com/aaronlehmann/b42a2eaf633fc949f93b

jhowardmsft commented:
@swernli - Stefan can you review from the Windows side please?

thaJeztah commented:
> The first time a version of the engine with these commits is started, it will migrate old graph metadata to the new format. This involves calculating content hashes for the existing data, but it does not move underlying graphdriver filesystem data. It doesn’t remove old graph metadata, so the migration process is not destructive.

Curious; will repeated upgrades / downgrades work? (e.g., Used 1.10 for testing, then continue on 1.9 and run 1.10 again)

aaronlehmann commented:
> Curious; will repeated upgrades / downgrades work? (e.g., Used 1.10 for testing, then continue on 1.9 and run 1.10 again)

New images built or pulled in 1.10 wouldn't be visible to 1.9. But the original images would still be there, unless you manually deleted some of them.

runcom commented on daemon/daemon.go:
```diff
@@ -1054,47 +1074,212 @@ func (daemon *Daemon) ImportImage(src, repo, tag, msg string, inConfig io.ReadCl
+		Architecture:    img.Architecture,
+		Os:              img.OS,
+		Size:            size,
+		//VirtualSize:   s.graph.GetParentsSize(image) + image.Size, // FIXME: document removal
```
should this be versioned? or can't it anymore?

tonistiigi replied:
`VirtualSize` as a concept has been removed as all images freely share all of there layers and no layer is unique to an image. So the value `Size` is now same that was previously `VirtualSize`.

I changed the PR so that `VirtualSize` isn't cleared any more but shows same data. That should provide correct data to older clients. New clients don't use this field any more. As this field isn't useful for new clients we can figure out a deprecation path and this can be done through API versions, but I think we can do that after this PR is merged.

runcom replied:
cool thanks @tonistiigi just making sure we won't forget this

swernli commented:
Reviewing for Windows...

thaJeztah commented:
ping @docker/maintainers please review and test this one. It's a huge change, but really exciting stuff.

also:
**please try to avoid merging large PRs while this PR isn't merged yet to avoid rebase hell**

jfrazelle commented:
\o/

On Thu, Nov 12, 2015 at 12:56 PM, Sebastiaan van Stijn <
notifications@github.com> wrote:

> ping @docker/maintainers
> <https://github.com/orgs/docker/teams/maintainers> please review and test
> this one. It's a huge change, but really exciting stuff.
>
> also:
> *please try to avoid merging large PRs while this PR isn't merged yet to
> avoid rebase hell*
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/17924#issuecomment-156230950>.
>

thaJeztah commented:
A difference I see with this one (probably expected, but just to verify)

Dockerfile
```Dockerfile
FROM ubuntu:14.04
ENV foo=bar
RUN echo hello > /foo
```
```bash
docker build -t foo .
```
Docker 1.9:
```bash
docker history foobar
IMAGE               CREATED             CREATED BY                                      SIZE                COMMENT
408581b4301b        7 seconds ago       /bin/sh -c echo hello > /foo                    6 B
012645df8321        7 seconds ago       /bin/sh -c #(nop) ENV foo=bar                   0 B
e9ae3c220b23        2 days ago          /bin/sh -c #(nop) CMD ["/bin/bash"]             0 B
a6785352b25c        2 days ago          /bin/sh -c sed -i 's/^#\s*\(deb.*universe\)$/   1.895 kB
0998bf8fb9e9        2 days ago          /bin/sh -c echo '#!/bin/sh' > /usr/sbin/polic   194.5 kB
0a85502c06c9        2 days ago          /bin/sh -c #(nop) ADD file:531ac3e55db4293b8f   187.7 MB
```

With this PR:
```bash
docker history foobar
IMAGE               CREATED             CREATED BY                                      SIZE                COMMENT
a12448c314a7        6 minutes ago       /bin/sh -c echo hello > /foo                    2.048 kB
e39f86d96664        6 minutes ago       /bin/sh -c #(nop) ENV foo=bar                   0 B
4b1e42b414f6        2 days ago          /bin/sh -c #(nop) CMD ["/bin/bash"]             1.024 kB
<missing>           2 days ago          /bin/sh -c sed -i 's/^#\s*\(deb.*universe\)$/   4.608 kB
<missing>           2 days ago          /bin/sh -c echo '#!/bin/sh' > /usr/sbin/polic   208.9 kB
<missing>           2 days ago          /bin/sh -c #(nop) ADD file:531ac3e55db4293b8f   196.8 MB
```

Differences: 

- sizes reported are different
- images from the parent image are reported as `<missing>` (I expect that to be by design, but just checking)

tonistiigi commented:
@thaJeztah The sizes are currently from tars so they have the added padding. We are exploring ways to revert this if possible. `<missing>` is shown if the parent image is not locally available. This is because you probably did a fresh pull and got a flat image without parents for ubuntu.

thaJeztah commented:
@tonistiigi correct, it was a fresh pull

swernli commented:
Currently, these changes break the use of alternate ID files in layers necessary for Windows containers to work.  I'm looking for a fix that can be applied to these changes to unblock us.

moxiegirl commented:
Not in doc review but this could have documentation impact that is pretty big depending on the implementation. So, engineering hours could be required in terms of reviewing at the least. I'd like to include this information early in the PR lifecycle.

* The image discussion recently added along with the storage driver material: https://github.com/docker/docker/blob/master/docs/userguide/storagedriver/imagesandcontainers.md  --- minor changes I think likely in graphics and in discussion of how images are built. (Not impact necessarily to choosing a storage driver).
* Re @thaJeztah's demo of the output of a command --> we have a lot of examples of that show image lists in the docs.  If the format of the output around images changes significantly we should make sure to update all the examples across all the projects --- these are things readers notice.

tianon commented:
> This is because you probably did a fresh pull and got a flat image without parents for ubuntu.

This part of this proposal seems really strange to me, and I'm trying to understand the rationale for making this change unilaterally for all images that get pushed/pulled -- right now, if I pull two images that have several layers with the same content and parents, they'll nicely share that layer (one of the main advertized advantages of Docker's layered image storage model) and I won't incur the space taken up by that layer twice.  With this change, overlapping layers won't matter anymore, and so if I `docker pull golang:1.4` and then `docker pull golang:1.5`, I get to incur the bandwidth and storage cost of both the `debian:jessie` they're based on _and_ the layers which are the same up until the point where the two `Dockerfile`s actually diverge?  https://imagelayers.io/?images=golang:1.4,golang:1.5 illustrates what I mean pretty well -- in this example, this behavior basically this means ~425MB of extra space taken up by the second image on my local system, plus whatever the compressed version of that is in bandwidth and time waiting for the pull.  Have I misunderstood the feature?  Having this as an _option_ sounds really neat, but applying it unilaterally by default with no opt-out feels like walking backwards IMO.

duglin commented:
if @tianon is correct about us no longer sharing layers then I'd prefer to not do this, we'd lose one of the biggest benefits that Docker offers the community.  Would it be possible for someone to explain why we can't keep the current layering/sharing but just use SHAs as the IDs instead of the random ones we use today?

cpuguy83 commented:
Layer sharing is still happening afaict, as evidenced by doing `docker pull golang:1.4 && docker pull golang:1.5 && docker pull debian:jessie`
The 1.5 pull will show several "already exists" layers, and the debian pull will be all "already exists".

tonistiigi commented:
@tianon @duglin All images can share their underlying layers freely. Flat image means that it doesn't need a parent to run, not some kind of isolation of layers. Sharing of layers is even better because we now identify individual layers based on content, not on arbitrary image IDs that they were connected to. So you only get new layers when layer content actually changes. As the title says everything is content addressable - so you don't get any duplication. Also we take out empty layers on build  that would have previously taken you out of sharing. Other changes include having a mapping to compressed blobsums, so we can actually detect layer sharing and skip a download even for different compressed blobs.

tianon commented:
Ahh, so we still store them separately, we just don't have a way to reference them individually?  I think that's still the part that's confusing me.  What situations cause layers to be `<missing>`, and would those layers (and that filesystem space) be reclaimable later since we would then have the metadata for the individual layers?  Are they still usable for build cache, etc?

tonistiigi commented:
@tianon Please see the design-doc about image config: https://gist.github.com/aaronlehmann/b42a2eaf633fc949f93b#new-image-config

In the flat model history is just array of text in the config that your builder can produce. There is no strict relationship between a history item and a layer. To become compatible with `docker history` UI we try to match them when generating that view. If the user has a parent image on disk that we can also match, we will show the parent ID as well.

If you build an image you will get a new image(still flat) for every dockerfile command. This means that if you run `docker history` on self-built image you should always have the image IDs there until you delete something. Also this allows us to have no modifications for the build cache.

tianon commented:
Right, I understand the way we're handling that part -- what I'm fuzzy on is the `rootfs` section of that doc (https://gist.github.com/aaronlehmann/b42a2eaf633fc949f93b#rootfs).  It _sounds_ (from the design spec) like we do still store each layer's contents separately on-disk, but from the discussion here it's fuzzy whether `docker pull golang:1.5` will give me a single `rootfs` blob, or whether I'll still get one `rootfs` blob for each layer.  If we're still getting individual layers that are just missing the "image metadata", then this sounds fine.

aaronlehmann commented:
> If we're still getting individual layers that are just missing the "image metadata", then this sounds fine.

Yes, that's the right way to think about it. The layers are shared whenever they have identical filesystem contents. When you pull something, you only get one image configuration instead of one for each layer.

tianon commented:
Cool, thanks for confirming! :+1:

tonistiigi commented:
@thaJeztah The layer size reporting has been switched back to file content based algorithm. This should fix the size differences in `docker history` output.

icecrime commented:
FYI I did an early review before PR was submitted, so count that as half an LGTM from me ;-) Awesome work @tonistiigi @aaronlehmann @dmcgowan :+1:

aidanhs commented:
This seems really nice. I have a few questions/comments:

1. The design doc says the parent key will be removed from image configuration and it will be "stored out-of-band". What exactly will the 'parent' of an image mean (maybe the id of the image used in `FROM`)? Where will it be used? It doesn't seem to me that the concept of a 'parent image' has much meaning any more.
2. It seems unfortunate that there's no way to build on top of a specific sub-chain. Taking https://github.com/docker/docker/pull/17924#issuecomment-156234758 as an example, there appears to be no way to build on top of the first `ADD` layer.
3. Related to 2, are you planning to add (or are open to adding) plumbing apis to do useful things like: a) export individual layers, b) import individual layers, c) create an image from an image configuration (list of layers etc). Now that layers are completely divorced from image configuration this seems like a particularly useful thing to do.
4. IMO you need big warnings in the changelog (and any other publicity) that the daemon will be totally unavailable until migration has happened and this may take a while if you have a lot of large images.
5. Why does `FSMetadataStore` create a bunch of files and directories (under `image/aufs/distribution`, i.e. `blobsum-{lookup,storage}/*`) rather than just maintaining a single json/sqlite file for lookup (e.g. `image/aufs/distribution.json`)? It uses a single mutex so race conditions aren't an issue.

aaronlehmann commented:
> The design doc says the parent key will be removed from image configuration and it will be "stored out-of-band". What exactly will the 'parent' of an image mean (maybe the id of the image used in FROM)? Where will it be used? It doesn't seem to me that the concept of a 'parent image' has much meaning any more.

The idea here is to allow an image to stand on its own, instead of requiring all the layers that were part of its build history. If the parent ID was part of the image configuration, the parent information would affect the hash of the configuration and therefore the image's ID.

> It seems unfortunate that there's no way to build on top of a specific sub-chain. Taking #17924 (comment) as an example, there appears to be no way to build on top of the first ADD layer.

This is correct in the context of push/pull, which only move around one image configuration. However, when you build something locally, it's still possible to build on top of a specific sub-chain.

> Related to 2, are you planning to add (or are open to adding) plumbing apis to do useful things like: a) export individual layers, b) import individual layers, c) create an image from an image configuration (list of layers etc). Now that layers are completely divorced from image configuration this seems like a particularly useful thing to do.

I imagine you're talking about HTTP APIs here, since these things are already available at the Go package level. I'm certainly open to exploring this, but what we had in mind longer-term was splitting the layer/image code out of the docker/docker repo, and building tools on top of it to do things like this.

> IMO you need big warnings in the changelog (and any other publicity) that the daemon will be totally unavailable until migration has happened and this may take a while if you have a lot of large images.

Yes, this needs to be made clear. We should also provide rough estimates of how long the migration takes as a function of the amount of layer data in the graph. BTW, since this is only hashing the data on disk, and not actually moving it, the migration goes relatively quickly.

> Why does FSMetadataStore create a bunch of files and directories (under image/aufs/distribution, i.e. blobsum-{lookup,storage}/*) rather than just maintaining a single json/sqlite file for lookup (e.g. image/aufs/distribution.json)? It uses a single mutex so race conditions aren't an issue.

My concern about a single JSON file is that it might grow quite large, and making any update to it will involve significant I/O and parsing overhead. sqlite may be a good approach, but I'm not sure whether sqlite use is encouraged (I believe there's a PR changing links to not use sqlite). I see that `boltdb` is already vendored, so that would be another option.

Since the metadata store is defined as an interface, it should be easy to experiment with other backends.

aidanhs commented:
Thanks for your response.

> The idea here is to allow an image to stand on its own, instead of requiring all the layers that were part of its build history. If the parent ID was part of the image configuration, the parent information would affect the hash of the configuration and therefore the image's ID.

I wasn't clear enough. I agree that all of this is good - my question is, why can't you do away with parent image ID completely (i.e. don't even store it 'out-of-band')? Is it just for a nicer output of `docker history`? I'm trying to figure out what people will now use it for/whether it's useful any more.

> This is correct in the context of push/pull, which only move around one image configuration. However, when you build something locally, it's still possible to build on top of a specific sub-chain.

Ok, given https://github.com/docker/docker/pull/17924#issuecomment-156234758, how would you build on top of the subchain consisting of just the `/bin/sh -c #(nop) ADD file:531ac3e55db4293b8f` layer?

> I imagine you're talking about HTTP APIs here, since these things are already available at the Go package level. I'm certainly open to exploring this, but what we had in mind longer-term was splitting the layer/image code out of the docker/docker repo, and building tools on top of it to do things like this.

Wouldn't you need a HTTP api level anyway? Any tool would need root permissions to fiddle around with docker graph directory. Happy to hear that this is planned though.

> Yes, this needs to be made clear. We should also provide rough estimates of how long the migration takes as a function of the amount of layer data in the graph. BTW, since this is only hashing the data on disk, and not actually moving it, the migration goes relatively quickly.

It'll also be much slower if you have many many small files, but that's an edge case. 
My /var/lib/docker is ~7GB (which is pretty tiny for my standards) and it took a number of mins. It's not too long, but people generally don't expect 5 mins downtime when restarting a service post-upgrade.

> My concern about a single JSON file is that it might grow quite large, and making any update to it will involve significant I/O and parsing overhead. sqlite may be a good approach, but I'm not sure whether sqlite use is encouraged (I believe there's a PR changing links to not use sqlite). I see that boltdb is already vendored, so that would be another option.

I don't mind, boltdb sounds good. At present you're just implementing your own a kv-store with the filesystem.

tonistiigi commented:
> why can't you do away with parent image ID completely (i.e. don't even store it 'out-of-band')?

When you do a build you will still get a separate image for every Dockerfile command. You can start/push/tag any of those images and they all make up a parent chain(view them with `docker images -a`). The chain is mainly used for builder cache. History also uses the parents to provide mapping if possible, but this probably isn't a high priority use-case.

thaJeztah commented:
@tonistiigi needs a rebase :cry: hope it's a simple one

tianon commented:
Just adding that I find it really useful while testing and developing
Dockerfiles to be able to jump into a container made from one of the
intermediate instructions, so thanks for making sure that's still included.
:+1: :heart:

aidanhs commented on distribution/pull_v1.go:
```diff
@@ -0,0 +1,447 @@
+		if layerID, err := p.v1IDService.Get(v1LayerID, p.repoInfo.Index.Name); err == nil {
+			// Does the layer actually exist
+			if l, err := p.config.LayerStore.Get(layerID); err == nil {
+				for j := i; i >= 0; i-- {
```
Pretty sure it needs to be `j >= 0`, `j--`.

aidanhs commented:
I pulled an image from a v1 registry, `rmi`d it and then tried to pull it again. Result:
```
Pulling repository company.internal:5000/image
8637682a033d: Error pulling image (1.0) from company.internal/image, too many non-empty layers in History section 00/v1/, too many non-empty layers in History section 
47d44cb6f252: Pulling metadata 
6a7b54515901: Pulling metadata 
[...]
Get https://company.internal:5000/v2/: EOF
endpoint does not support v2 API
unable to ping registry endpoint https://company.internal:5000/v0/
v2 ping attempt failed with error: Get https://company.internal:5000/v2/: EOF
 v1 ping attempt failed with error: Get https://company.internal:5000/v1/_ping: EOF
Error pulling image (1.0) from company.internal:5000/image, too many non-empty layers in History section
```

it may be related to the comment I've just made, not yet sure.

aaronlehmann commented:
@aidanhs: This was indeed caused by the bad loop conditions. Thank you for pinpointing it. I've updated the commit to fix the problem. This didn't show up in my original testing because the problem only appears if you already have some, but not all, of the layers for a v1 image.

dmcgowan committed:
Add layer store

Layer store manages read-only and read-write layers on a union file system.
Read only layers are always referenced by content addresses.
Read-write layer identifiers are handled by the caller but upon registering
its difference, the committed read-only layer will be referenced by content
hash.

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)
Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

Commit:
Add tag store

The tag store associates tags and digests with image IDs. This
functionality used to be part of graph package. This commit splits it
off into a self-contained package with a simple interface.

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

tonistiigi committed:
Add image store

The image store abstracts image handling. It keeps track of the
available images, and makes it possible to delete existing images or
register new ones. The image store holds references to the underlying
layers for each image.

The image/v1 package provides compatibility functions for interoperating
with older (non-content-addressable) image structures.

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

Commit:
Add distribution package

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

tonistiigi committed:
Add migratev1 package

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

tonistiigi committed:
Update daemon and docker core to use new content addressable storage

Add distribution package for managing pulls and pushes. This is based on
the old code in the graph package, with major changes to work with the
new image/layer model.

Add v1 migration code.

Update registry, api/*, and daemon packages to use the reference
package's types where applicable.

Update daemon package to use image/layer/tag stores instead of the graph
package

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>
Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

tonistiigi committed:
Remove graph package

The graph package is no longer used for tag, image, or layer storage.

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

tonistiigi commented:
@swernli I managed to get the windows baselayer support working. Tested pull/push/run/build/history/save/load in TP3. Patched tar-split with a crazy hack atm, you can have a better solution with some changes to your storage layer. My suggestion is that you should probably only put the files that never change in the tar and generate other metadata. 

Base layer in config is now text based like we discussed, so if you tested my earlier patch that used 64 byte hex you may need to clear your graph.

```
"rootfs":{
  "type":"layers+base",
  "diff_ids":["sha256:c438aa1198fb8574036eeae2ff3864b3ab06a8b1b6dd707c82a199413ef75e4f",...],
  "base_layer":"CN=Microsoft_WindowsServerCore_10.0.10514.0"
}
```

Here is the change without squash: https://github.com/docker/docker/commit/d537d85c232e6c2db87ec4aa52597a3762182455

vbatts commented:
OHMAN

swernli commented:
@tonistiigi I was able to pull up the thread I had from last time we encountered problems with tarsplit.  Unfortunately, our conclusion from last time was that tarsplit wasn't really compatible with the way Windows layers are exported and imported, since the content of the layer changes significantly as part of that process.  We can discuss more on IRC, but it may be that we'll need to rely on a hack for a while there...

crosbymichael commented on api/client/build.go:
```diff
@@ -593,29 +590,39 @@ func rewriteDockerfileFrom(dockerfileName string, translator func(string, regist
-			if !ref.HasDigest() && isTrusted() {
-				trustedRef, err := translator(repo, ref)
+			if !digested && isTrusted() {
+				trustedRef, err := translator(ref.(reference.NamedTagged))
```
This could panic if you do the type cast this way, do you have to be safer with this type?

dmcgowan replied:
The above switch statement already guarantees that this type `reference.NamedTagged` when it is not digested

aaronlehmann replied:
The type switch above should take care of this. The cases are:

- `reference.Digested`: We set `digested = true` and this line is not reached.
- `reference.Tagged`: The type returned by `ParseNamed` is `Named`. `NamedTagged` is a combination of `Named` and `Tagged` so if the cast to `Tagged` succeeds, the cast to `NamedTagged` must also succeed. We could replace this case label with `reference.NamedTagged` to make this more explicit.
- `default`: `ref` gets replaced with the result of `reference.WithTag`, which returns a `NamedTagged`.

crosbymichael replied:
so do the type case in the switch so that you don't have to cast it 2 more times outside of the switch?

crosbymichael replied:
if you don't know u can get the type like this: 

https://golang.org/doc/effective_go.html#type_switch

crosbymichael commented on api/client/pull.go:
```diff
@@ -23,9 +23,25 @@ func (cli *DockerCli) CmdPull(args ...string) error {
+		tag = x.Digest().String()
+	case reference.Tagged:
+		tag = x.Tag()
+	}
```
what about a default case?

aaronlehmann replied:
> what about a default case?

Changed this to use a default case instead of if/else right after the switch.

aidanhs commented on distribution/registry.go:
```diff
@@ -59,9 +57,9 @@ func newV2Repository(repoInfo *registry.RepositoryInfo, endpoint registry.APIEnd
+		Timeout:   5 * time.Second,
 	}
-	endpointStr := strings.TrimRight(endpoint.URL, "/") + "/v2/"
+	endpointStr := endpoint.URL + "/v2/"
```
You've reverted #17203?

aaronlehmann replied:
> You've reverted #17203?

Yes, unintentionally. Since the package name has changed, these kinds of changes need to be handled manually. This one seems to have slipped through.

I've added the missing change. Thanks for the careful review!

aidanhs commented on daemon/images.go:
```diff
@@ -0,0 +1,163 @@
+			if err != nil {
+				return nil, err
+			}
+			layer.ReleaseAndLog(daemon.layerStore, l)
```
Use the code arrangement https://github.com/docker/docker/pull/17924/files#diff-1a1f3e7ad9b1d7584e2d3e7d0c4c3db9R1192 so you release even if there's an error getting size?

aidanhs commented on daemon/daemon.go:
```diff
@@ -1051,47 +1070,211 @@ func (daemon *Daemon) ImportImage(src, repo, tag, msg string, inConfig io.ReadCl
+			return nil, err
+		}
+		defer layer.ReleaseAndLog(daemon.layerStore, l)
+		size, _ = l.Size()
```
Why no error checking?

crosbymichael commented:
testing looks good on upgrade and push/pull

LGTM

aidanhs commented:
Made a PR onto the branch - https://github.com/aaronlehmann/docker/pull/2
Additionally, why is `func ReleaseAndLog(ls Store, l Layer) {` not a method on `Store`? (I would make this a comment on the actual piece of code, but the diff is too big :( )

vbatts commented:
I'm still kicking this changeset around, but so far it's looking good to me

calavera commented:
I've seen some `panic` calls in `layer/layer_store.go`. Are those on purpose? they look very dangerous.

cpuguy83 commented:
yeah, I saw those panics as well... I traced them and really they should be panics if it got there as it really shouldn't ever... it's still a bit weird and scary.

tonistiigi commented:
@calavera These should be error cases that can only happen if there is a internal algorithm bug in layerStore. We could just ignore them (and let code crash in random point). cc @dmcgowan

LK4D4 commented:
@tonistiigi @aaronlehmann Are we waiting on merge @aidanhs PRs?

aaronlehmann commented:
@LK4D4: No need to wait for them. They are cleanups that can easily be submitted as followup PRs.

aidanhs commented:
I would like to repeat the suggestion to use some kind of real db rather than throwing loads of files onto disk, as once it's merged this PR is merged I can't see it being changed (you'll need a second migration process).

If it's agreed that this is desirable then I can look at doing it, but I think there should be a decision one way or another.

aaronlehmann commented:
I'm fine with the idea of using a database, but I don't think that the current approach is problematic. In general, it creates one or two files per layer, which is not very many files compared to the number of files in the layers themselves. Before this PR, there were similar files in the `graph` directory, such as `checksum`.

Also, I think it would be relatively easy to transition to a different storage scheme, because the files in the `distribution` directory just cache information about pushes and pulls. The worst case from changing the way they are stored is some unnecessary pushes and pulls for people who have run the intermediate version. If we switch to a database, I think the right way to do it is in a followup PR pre-1.10.

LK4D4 commented:
Actually it's pretty sad, that migration process is so slow :/ Seems like we can't do anything.

aidanhs commented:
I'm hoping I don't need to worry about lines like
```
ERRO[0012] migration failed for 5630927819e7c9439478f0c762993fdf044d0921b4d210acb1bd6ae6f36ab6e0, err: open /var/lib/docker/aufs/diff/5630927819e7c9439478f0c762993fdf04
4d0921b4d210acb1bd6ae6f36ab6e0/etc/group: no such file or directory 
ERRO[0012] migration failed for ee6677401abbf3578319ed0268d6cbcfd57a312289a16868b047e835f4fe55d4, err: open /var/lib/docker/aufs/diff/5630927819e7c9439478f0c762993fdf04
4d0921b4d210acb1bd6ae6f36ab6e0/etc/group: no such file or directory 
```
?

tonistiigi commented:
@aidanhs can you send me more info about that case, the `/var/lib/docker/graph/5630927819e7c9439478f0c762993fdf044d0921b4d210acb1bd6ae6f36ab6e0/tar-data.json.gz` file and `ls -l /var/lib/docker/aufs/diff/5630927819e7c9439478f0c762993fdf044d0921b4d210acb1bd6ae6f36ab6e0/etc/group`. Most likely something your tar-split file didn't have required data(that meant that you should get same error if you pushed that image with old version). If this is the case we can generate new `tar-split` file on error, but we should also figure out what causes these errors(one of them is #15643).

aidanhs commented:
@tonistiigi sorry, I unfortunately can't share those files because I nuked my docker data directory about 5 mins after the error and now can't reproduct. It had an accumulation of junk from other dev branches and compiling on master and docker versions back a fair way. I'll raise an issue if I see it happen again after this PR is merged.

cpuguy83 commented:
Migration is _really_ slow.  Worried that systemd will kill daemons in the middle of migration because it doesn't appear started yet, as it has a 10s timeout.
Starting my daemon with a handful of small images (debian:jessie, redis, ubuntu, busybox) took 12s.
Wonder if we can parallelize some of that work.

LK4D4 commented:
@cpuguy83 was 5 minutes for me :)

aaronlehmann commented:
@cpuguy83: I've addressed your latest comments in three new commits. These will be squashed before merge.

cpuguy83 commented:
There's a number of `TODO` and `FIXME` comments. I'm not sure if these are meant to be resolved before we're ready to merge or not.

cpuguy83 commented:
I have a feeling migration is going to cause a lot of problems, particularly on hosts with large images (and as such longer migration times).

I'm assuming the majority of time during migration is spent summing layers?
I wonder if we can refer to these by v1 ID's internally and kick off a job to a worker queue to get the sums. While sums are not populated the image can't be pushed/exported or used by content addressable ID.
This should also be an improvement for `commit` we don't really need the checksum until it needs to get pushed/exported (or do we???).

thaJeztah commented:
@cpuguy83 we were actually discussing some option last week, e.g. calculate the checksums in a queue, or even using a separate migration tool, so that the migration can be performed before upgrading docker. General opinion was that a solution has to be found *before* the final 1.10 release, but possibly as a separate PR

LK4D4 commented:
My opinion, that it should be tried in this PR. Because what if it'll be very hard or impossible to do? Revert will be not so easy.

aaronlehmann commented:
> There's a number of TODO and FIXME comments. I'm not sure if these are meant to be resolved before we're ready to merge or not.

We aren't planning to address any of these before the merge. Many are just from moved code.

crosbymichael commented:
For the download needs we can create a migration tool to perform the migration live on a running daemon before upgrade.  

 @tonistiigi is working on making the migration work concurrent and there is around a 25% performance boost but for others I think the migration tool will be the best thing to do.  

This can be merged as is and we can have smaller prs for optimizations and a migration tool.  Its not like @tonistiigi and @aaronlehmann are going to stop working on this after this pr is merged.

calavera commented:
LGTM.

I agree with @crosbymichael that we should not try to land all the migration work in this PR. Otherwise we'll end up merging this very close to the release with very little time to more exhaustive testing in real environments.

LK4D4 commented:
@tonistiigi said that migration-tool is not an option. I want to hear plan for migration stuff at least and not "okay, all need to have 25 minutes downtime without user-facing benefits" at release time.

cpuguy83 commented:
I'll have to side with @LK4D4 on this. The patch isn't ready if we don't have a safe way to migrate.

tonistiigi commented:
@LK4D4 Sorry, some miscommunication here. Migration-tool is definitely an option and work with parallel migration is already designed with that possibility in mind.

LK4D4 commented:
@tonistiigi cool :) then we have enough lgtms and you can merge PR :P

thaJeztah commented:
do we need docs changes as part of this PR, or will we address those in a follow-up? We discussed that there's some change in behavior, and we may need to address those in the docs

crosbymichael commented:
there should not be any change of behavior, this is all internal refactoring and there is no way we are going through a docs review in this pr

tonistiigi commented:
There are some docs changes needed(mostly for clarity, and some command outputs). They are already being worked on https://github.com/tonistiigi/docker/commit/fc5ba5e3832f748b34c69255cc0d9be69c2e719 and will be posted/reviewed with another PR.

vbatts commented:
Including a notice/FYI that anyone daring enough to crawl /var/lib/docker as an API, will likely be broken by this.

I have only been able to go through behavioral reviews of this PR, not nity grity code review. It still LGTM.

calavera commented:
Thank you all for your feedback. I'm merging it :metal:

bracki commented:
Does this change have any effects on the build cache? So can the build cache reuse pulled images now as well? Or would that be a separate issue? Could you elaborate on that @tonistiigi?

tonistiigi commented:
@bracki Build cache should work as expected. Cache is for remembering the data from previous builds. It isn't based on checksums but a comparison of image configurations.

dbdd4us commented:
> I imagine you're talking about HTTP APIs here, since these things are already available at the Go package level. I'm certainly open to exploring this, but what we had in mind longer-term was splitting the layer/image code out of the docker/docker repo, and building tools on top of it to do things like this.

Could you tell me what's the name of the package? Thank you.

dbdd4us commented:
> I imagine you're talking about HTTP APIs here, since these things are already available at the Go package level. I'm certainly open to exploring this, but what we had in mind longer-term was splitting the layer/image code out of the docker/docker repo, and building tools on top of it to do things like this.

@aaronlehmann  Could you tell me what's the name of the package? Thank you.

aaronlehmann commented:
@dbdd4us: The packages I was talking about are called `image` and `layer`. They are in the top-level directory of this repository.

dbdd4us commented:
> The packages I was talking about are called image and layer. They are in the top-level directory of this repository.
@aaronlehmann Thanks.

dbdd4us commented:
@tonistiigi Hello, Is there any tools that I can used to calculate the content-hash of a tar file?  
I have tried `tar -xOf layer.tar | sha256sum`, but the hash value I got is different from the hash value that calculated from docker daemon.
Or is there anything special in the go tar-split package?

stevvooe commented:
@dbdd4us The content hash of a layer is the hash of the compressed content, as sent to the registry.

tonistiigi commented:
@stevvooe layer hashes in image config are the hashes of uncompressed content.
@dbdd4us I'm not sure what the `layer.tar` is in your example. Checksum is for all the bytes in tar, there is no extraction step involved. Easiest to get those tars is with `docker save`. Note that these do not match the image IDs that are checksums of image config + layer checksums. Make sure to also read the design doc in first post.

ayanamist commented on layer/layer_store.go:
```diff
@@ -0,0 +1,649 @@
+	mountID := name
+	if runtime.GOOS != "windows" {
+		// windows has issues if container ID doesn't match mount ID
+		mountID = stringid.GenerateRandomID()
```
Hi, i'm investigating moby source code, when i read this, i doubt if there is any reason that moby do not always use `name` regard of windows or not. Is there any consideration not commented here?
